include/mfmg/adapters.hpp

```diff
@@ -0,0 +1,102 @@
+/*************************************************************************
+ * Copyright (c) 2017 by the mfmg authors                                *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef MFMG_ADAPTERS_HPP
+#define MFMG_ADAPTERS_HPP
+
+#include "amge_device.cuh"
+#include "amge_host.hpp"
+
+#include <deal.II/lac/trilinos_linear_operator.h>
+
+namespace mfmg
+{
+
+struct Cuda
+{
+};
+struct Host
+{
+};
+
+template <class Operator, class MultiVector, class Mesh, class Device>
+class Adapter
+{
+public:
+  static std::shared_ptr<Operator>
+  build_restrictor(MPI_Comm comm, const Operator &op, const Mesh &mesh);
+};
+
+template <int dim>
+class Adapter<
+    dealii::LinearOperator<
+        dealii::Vector<double>, dealii::Vector<double>,
+        dealii::TrilinosWrappers::internal::LinearOperator::TrilinosPayload>,
+    dealii::Vector<double>, dealii::DoFHandler<dim>, Host>
+{
+private:
+  using Operator = dealii::LinearOperator<
+      dealii::Vector<double>, dealii::Vector<double>,
```

Comment:
I am not sure this is what you want. `dealii::Vector` is serial. If you want a distributed vector, you need to use `dealii::LinearAlgebra::distributed::Vector` or `dealii::TrilinosWrappers::MPI::Vector`

Reply:
I'm getting lost in all the dealii types :( But you are right.